### PR TITLE
[MM-56032] Add tests for data_retention endpoints

### PR DESCRIPTION
**Pull request by fume4mattermost**
#### Summary

Added unit tests for the functions in data data_retention.go

#### Ticket Link

https://mattermost.atlassian.net/browse/MM-56032

#### Release Note

```release-note

```

### CODE REVIEW

**Comment by jwilander** [comment]
@fume4mattermost looks like some [Go linting isn't passing](https://github.com/mattermost/mattermost/actions/runs/10993755993/job/30565870020?pr=28283) - can you address that?

**Review by hanzei — CHANGES_REQUESTED** [review]
https://github.com/mattermost/mattermost/pull/28283#issuecomment-2375119767

**Review by hanzei — CHANGES_REQUESTED** [review]
@fume4mattermost  Some of the tests are not passing. Can you please take a look?

**Comment by jwilander** [comment]
@fume4mattermost seeing the following lint errors:

```
Error: channels/api4/data_retention_test.go:90:19: Error return value of `th.Client.Logout` is not checked (errcheck)
		th.Client.Logout(context.Background())
		                ^
Error: channels/api4/data_retention_test.go:224:19: Error return value of `th.Client.Logout` is not checked (errcheck)
		th.Client.Logout(context.Background())
		                ^
Error: channels/api4/data_retention_test.go:319:19: Error return value of `th.Client.Logout` is not checked (errcheck)
		th.Client.Logout(context.Background())
```

**Comment by BenCookie95** [comment]
@fume4mattermost Looks like there are still a few more lint errors:
```
Error: channels/api4/data_retention_test.go:435:19: Error return value of `th.Client.Logout` is not checked (errcheck)
		th.Client.Logout(context.Background())
		                ^
Error: channels/api4/data_retention_test.go:512:19: Error return value of `th.Client.Logout` is not checked (errcheck)
		th.Client.Logout(context.Background())
		                ^
Error: channels/api4/data_retention_test.go:602:19: Error return value of `th.Client.Logout` is not checked (errcheck)
		th.Client.Logout(context.Background())
```

**Review by BenCookie95 — CHANGES_REQUESTED** [review]
Lint errors from https://github.com/mattermost/mattermost/pull/28283#issuecomment-2393950720

**Comment by fume4mattermost** [comment]
Hey @BenCookie95. The linting errors are now fixed. Can you look at this again?

**Review by BenCookie95 — APPROVED** [review]
Thank you @fume4mattermost, these tests are extremely helpful 🚀

File: server/channels/api4/data_retention_test.go [hunk 1]
```diff
@@ -18,3 +24,1240 @@ func TestDataRetentionGetPolicy(t *testing.T) {
...
+			},
+			ChannelCount: 1,
+			TeamCount:    1,
+		},
+	}
+	samplePolicyList := &model.RetentionPolicyWithTeamAndChannelCountsList{
+		Policies:   samplePolicies,
+		TotalCount: int64(len(samplePolicies)),
+	}
+	mockDataRetentionInterface.On("GetPolicies", mock.Anything, mock.Anything).Return(samplePolicyList, nil)
```
Comment (hanzei) [inline server/channels/api4/data_retention_test.go:59]: Can we be a bit stricter here and use something else than `mock.Anything`?

File: server/channels/api4/data_retention_test.go [hunk 2]
```diff
@@ -18,3 +24,1240 @@ func TestDataRetentionGetPolicy(t *testing.T) {
...
+
+	// Set the mock on the app
+	th.App.Srv().Channels().DataRetention = mockDataRetentionInterface
+
+	t.Run("GetPolicies_Success", func(t *testing.T) {
+		th.LoginSystemAdmin()
```
Comment (hanzei) [inline server/channels/api4/data_retention_test.go:65]: Is that call really required?

File: server/channels/api4/data_retention_test.go [hunk 3]
```diff
@@ -18,3 +24,1240 @@ func TestDataRetentionGetPolicy(t *testing.T) {
...
+		require.NotNil(t, policies, "Policies should not be nil")
+		require.NotEmpty(t, policies.Policies, "Policies should not be empty")
+		require.Len(t, policies.Policies, 1, "Should return 1 policy")
+		assert.Equal(t, "sample_policy_id", policies.Policies[0].ID, "Policy ID should match")
+		assert.Equal(t, int64(1), policies.Policies[0].ChannelCount, "ChannelCount should be 1")
+		assert.Equal(t, int64(1), policies.Policies[0].TeamCount, "TeamCount should be 1")
+	})
+
+	t.Run("GetPolicies_Forbidden", func(t *testing.T) {
+		th.LoginBasic()
```
Comment (hanzei) [inline server/channels/api4/data_retention_test.go:79]: Is that call really required?

File: server/channels/api4/data_retention_test.go [hunk 4]
```diff
@@ -18,3 +24,1240 @@ func TestDataRetentionGetPolicy(t *testing.T) {
...
+		th.RemovePermissionFromRole(model.PermissionSysconsoleReadComplianceDataRetentionPolicy.Id, model.SystemUserRoleId)
+
+		policies, resp, err := th.Client.GetDataRetentionPolicies(context.Background(), 0, 100)
+		require.Error(t, err)
+		CheckForbiddenStatus(t, resp)
+		assert.Nil(t, policies, "Policies should be nil on forbidden access")
+	})
+
+	t.Run("GetPolicies_Unauthorized", func(t *testing.T) {
+		_, err := th.Client.Logout(context.Background())
```
Comment (hanzei) [inline server/channels/api4/data_retention_test.go:90]: Can you please also check the first return variable? Thanks!

File: server/channels/api4/data_retention_test.go [hunk 5]
```diff
@@ -18,3 +24,1240 @@ func TestDataRetentionGetPolicy(t *testing.T) {
 	require.Error(t, err)
 	CheckNotImplementedStatus(t, resp)
 }
+func TestGetPolicies(t *testing.T) {
+	th := Setup(t).InitBasic()
+	defer th.TearDown()
+
+	// Set up a test license with Data Retention enabled
+	th.App.Srv().SetLicense(model.NewTestLicense("data_retention"))
```
Comment (hanzei) [inline server/channels/api4/data_retention_test.go:32]: `SetLicense` returns a boolean. Please check that it is `true`.

File: server/channels/api4/data_retention_test.go [hunk 6]
```diff
@@ -18,3 +24,1240 @@ func TestDataRetentionGetPolicy(t *testing.T) {
...
+	th.App.UpdateConfig(func(cfg *model.Config) {
+		*cfg.DataRetentionSettings.EnableMessageDeletion = true
+		*cfg.DataRetentionSettings.EnableFileDeletion = true
+	})
+
+	// Create and set up the mock
+	mockDataRetentionInterface := &mocks.DataRetentionInterface{}
+
+	// Set up the mock to return a count of 3 policies
+	mockDataRetentionInterface.On("GetPoliciesCount").Return(int64(3), (*model.AppError)(nil))
```
Comment (hanzei) [inline server/channels/api4/data_retention_test.go:118]: Is returning `nil` enough?

File: server/channels/api4/data_retention_test.go [hunk 7]
```diff
@@ -18,3 +24,1240 @@ func TestDataRetentionGetPolicy(t *testing.T) {
...
+
+	// Set the mock on the app
+	th.App.Channels().DataRetention = (mockDataRetentionInterface)
```
Comment (hanzei) [inline server/channels/api4/data_retention_test.go:121]: Why are the braces needed?

File: server/channels/api4/data_retention_test.go [hunk 8]
```diff
@@ -18,3 +24,1240 @@ func TestDataRetentionGetPolicy(t *testing.T) {
...
+	t.Run("get policies count without permissions", func(t *testing.T) {
+		th.LoginBasic()
+		_, resp, err := th.Client.GetDataRetentionPoliciesCount(context.Background())
```
Comment (hanzei) [inline server/channels/api4/data_retention_test.go:124]: Please also check the first return variable.

File: server/channels/api4/data_retention_test.go [hunk 9]
```diff
@@ -18,3 +24,1240 @@ func TestDataRetentionGetPolicy(t *testing.T) {
...
+
+	t.Run("GetPolicy_Success", func(t *testing.T) {
+		th.LoginSystemAdmin()
+
+		policy, resp, err := th.SystemAdminClient.GetDataRetentionPolicyByID(context.Background(), validPolicyId)
+		if err != nil {
+			t.Logf("Error: %v", err)
+			t.Logf("Response Status Code: %d", resp.StatusCode)
+			t.Logf("Response Error: %v", resp.Etag)
+		}
```
Comment (hanzei) [inline server/channels/api4/data_retention_test.go:193]: Please remove this block

File: server/channels/api4/data_retention_test.go [hunk 10]
```diff
@@ -18,3 +24,1240 @@ func TestDataRetentionGetPolicy(t *testing.T) {
...
+		th.RemovePermissionFromRole(model.PermissionSysconsoleReadComplianceDataRetentionPolicy.Id, model.SystemUserRoleId)
+
+		policy, resp, err := th.Client.GetDataRetentionPolicyByID(context.Background(), validPolicyId)
+		require.Error(t, err)
+		CheckForbiddenStatus(t, resp)
+		assert.Nil(t, policy, "Policy should be nil on forbidden access")
+	})
+
+	t.Run("GetPolicy_Unauthorized", func(t *testing.T) {
+		_, err := th.Client.Logout(context.Background())
```
Comment (hanzei) [inline server/channels/api4/data_retention_test.go:225]: Please also check the first return variable

File: server/channels/api4/data_retention_test.go [hunk 11]
```diff
@@ -18,3 +24,1240 @@ func TestDataRetentionGetPolicy(t *testing.T) {
...
+		}
+
+		policy, resp, err := th.Client.CreateDataRetentionPolicy(context.Background(), policyToCreate)
+		require.Error(t, err)
+		CheckForbiddenStatus(t, resp)
+		assert.Nil(t, policy, "Policy should be nil on forbidden access")
+	})
+
+	t.Run("CreatePolicy_Unauthorized", func(t *testing.T) {
+		_, err := th.Client.Logout(context.Background())
```
Comment (hanzei) [inline server/channels/api4/data_retention_test.go:321]: Please also check the first return variable

**Review by hanzei — CHANGES_REQUESTED** [review]
Can you please double-check if the `th.Login` calls are needed?

Secondly, the sub tests should not duplicate the test name of the parent test. Please update them.

**Comment by jwilander** [comment]
@fume4mattermost looks the the CI lint check is failing - can you take a look?

File: server/channels/api4/data_retention_test.go [hunk 12]
```diff
@@ -18,3 +24,1220 @@ func TestDataRetentionGetPolicy(t *testing.T) {
...
+		require.Len(t, policies.Policies, 1, "Should return 1 policy")
+		assert.Equal(t, "sample_policy_id", policies.Policies[0].ID, "Policy ID should match")
+		assert.Equal(t, int64(1), policies.Policies[0].ChannelCount, "ChannelCount should be 1")
+		assert.Equal(t, int64(1), policies.Policies[0].TeamCount, "TeamCount should be 1")
+	})
+
+	t.Run("with different page and per_page", func(t *testing.T) {
+		page := 1
+		perPage := 50
+		t.Logf("Calling GetDataRetentionPolicies with page=%d, perPage=%d", page, perPage)
```
Comment (hanzei) [inline server/channels/api4/data_retention_test.go:84]: Can you please remove all calls to `t.Logf` from this file? Thanks.

**Review by hanzei — APPROVED** [review]
LGTM :+1:

**Review by hanzei — CHANGES_REQUESTED** [review]
Blocking on https://github.com/mattermost/mattermost/pull/28283#discussion_r1816307524

**Review by hanzei — APPROVED** [review]

**Pull request merged**